services/src/main/java/org/fao/geonet/api/es/queryrewrite/FavouritesListESQueryRewriter.java

```diff
@@ -0,0 +1,299 @@
+
+    @Autowired
+    FavouriteMetadataListRepository userMetadataSelectionListRepo;
+
+    @Autowired
+    UserRepository userRepository;
+
+    /**
+     * rewrites an elasticsearch query.
+     *
+     * If there is an exception, it is logged.  However, this will return with an unmodified query.
+     *
+     * @param httpSession - used to get the current User (for security)
+     * @param request - used to get the current http session (for security)
+     * @param jsonESQuery - original query
+     * @return
+     */
+    @Override
+    public String rewriteQuery(HttpSession httpSession, HttpServletRequest request, String jsonESQuery)  {
+        try {
+            ObjectMapper mapper = new ObjectMapper();
+            JsonNode root = mapper.readTree(jsonESQuery);
+
+            ArrayNode queryItems = (ArrayNode) root.get("query").get("function_score").get("query").get("bool").get("must");
+            int index = -1;
+            for (JsonNode node: queryItems) {
+                index++;
+                if (!node.isObject()) {
+                    continue;
+                }
+                ObjectNode objectNode = (ObjectNode) node;
+                if (!objectNode.has("terms")) {
+                    continue;
+                }
+                ObjectNode termsNode = (ObjectNode) objectNode.get("terms");
+                if (!termsNode.has("favouritesList")) {
+                    continue;
+                }
+                int selectionId =  ((IntNode)termsNode.get("favouritesList").get(0)).asInt();
+                ObjectNode queryString = replacement(selectionId,mapper, httpSession, request);
+                queryItems.remove(index);
+                queryItems.add(queryString);
+                fixsort((ObjectNode) root);
+                String result = mapper.writeValueAsString(root);
+                return result;
+            }
+            return jsonESQuery;
+        }
+        catch (Exception e) {
+            //don't modify
+            Log.debug(Geonet.SEARCH_ENGINE, "FavouritesListESQueryRewriter",e);
+            return jsonESQuery;
+        }
+    }
+
+    private void fixsort(ObjectNode root) {
+        if (!root.has("sort")) {
+            root.putArray("sort").add("_id");
+        }
+    }
+
+
+    /**
+     * constructs the
+     *
+     *  {
+     *    query:
+     *     {
+     *         query_string: "_id(uuid1) OR _id(uuid2) ..."
+     *     }
+     *  }
+     *
+     *  that will replace the
+     *
+     *           {
+     *                 "terms": {
+     *                   "favouritesList": [
+     *                        104
+     *                    ]
+     *                }
+     *            }
+     *
+     *   This checks security (i.e. does the user/session have read permission on the favourites list?)
+     */
+    ObjectNode replacement(int favouritesListId, ObjectMapper mapper, HttpSession httpSession, HttpServletRequest request) throws Exception {
```

Comment:
If the visibility of this method is package private just for the tests, could we annotate it with `@VisibleForTesting` to have clear that it's a private method?
```suggestion
    @VisibleForTesting
    ObjectNode replacement(int favouritesListId, ObjectMapper mapper, HttpSession httpSession, HttpServletRequest request) throws Exception {
```